apisix/discovery/k8s.lua

```diff
@@ -0,0 +1,472 @@
+end
+
+local _M = {
+    version = 0.01
+}
+
+function _M.nodes(service_name)
+    local pattern = "^(.*):(.*)$"
+    local match, _ = ngx.re.match(service_name, pattern, "jiao")
+    if not match then
+        core.log.info("get unexpected upstream service_name:　", service_name)
+        return nil
+    end
+
+    local endpoint_key = match[1]
+    local endpoint_port = match[2]
+    local endpoint_version, _, _ = endpoints_shared:get_stale(endpoint_key .. "#version")
+    if not endpoint_version then
+        core.log.info("get empty endpoint version from discovery DICT ", endpoint_key)
+        return nil
+    end
+    return endpoint_lrucache(service_name, endpoint_version, create_endpoint_lrucache, endpoint_key, endpoint_port)
+end
+
+local function fill_pending_resources()
+    pending_resources = core.table.new(1, 0)
+
+    pending_resources[1] = {
+
+        group = "",
+        version = "v1",
+        kind = "Endpoints",
+        listKind = "EndpointsList",
+        plural = "endpoints",
+
+        newest_resource_version = "",
+
+        label_selector = function()
+            return ""
+        end,
+
+        list_path = function(self)
+            return "/api/v1/endpoints"
+        end,
+
+        list_query = function(self, continue)
+            if continue == nil or continue == "" then
+                return "limit=32"
+            else
+                return "limit=32&continue=" .. continue
+            end
+        end,
+
+        watch_path = function(self)
+            return "/api/v1/endpoints"
+        end,
+
+        watch_query = function(self, timeout)
+            return string.format("watch=true&allowWatchBookmarks=true&timeoutSeconds=%d&resourceVersion=%s", timeout,
+                    self.newest_resource_version)
+        end,
+
+        pre_list_callback = function(self)
+            self.newest_resource_version = "0"
+            endpoints_shared:flush_all()
+        end,
+
+        post_list_callback = function(self)
+            endpoints_shared:flush_expired()
+        end,
+
+        added_callback = function(self, object)
+            on_endpoint_added(object)
+        end,
+
+        modified_callback = function(self, object)
+            on_endpoint_modified(object)
+        end,
+
+        deleted_callback = function(self, object)
+            on_endpoint_deleted(object)
+        end,
+
+        event_dispatch = function(self, event, object)
+            if event == "DELETED" or object.deletionTimestamp ~= nil then
+                self:deleted_callback(object)
+                return
+            end
+
+            if event == "ADDED" then
+                self:added_callback(object)
+            elseif event == "MODIFIED" then
+                self:modified_callback(object)
+            end
+        end,
+    }
+end
+
+function _M.init_worker()
+    if process.type() ~= "privileged agent" then
+        return
+    end
+
+    apiserver_host = os.getenv("KUBERNETES_SERVICE_HOST")
```

Comment:
refer to other discovery modules, we may define these variables in the config.yaml.

Reply:
k8s discovery is expected to run in a k8s pod,
In this case, I think setting "KUBERNETES_SERVICE_HOST" and "KUBERNETES_SERVICE_PORT" via config.yaml is meaningless.

Reply:
I don't think so, APISIX itself isn't necessary to be run inside a Pod, but can be deployed outside of the Kubernetes cluster.

Reply:
Yes, APISIX can deployed outside the k8s cluster, but APISIX can't connect to the pod ip address.

Reply:
> Yes, APISIX can deployed outside the k8s cluster , but can not connect to the pod ip address

This is decided by how the Kubernetes networking designs.

Anyway, I think configuration items and env can be supported at the same time, env can be ineffective if configuration items are absent.

Reply:
Ok, I will do something to support this.

Reply:
When the client is located outside the cluster, there are usually three ways to authenticate with the apiserver
1. MTLS
2. Service Account Token
3. Http Basic Auth

But based on two points:
+ Openrety Cosocket does not currently support mtls.
+ Http Basic is not very secure

I think it only needs to support service account token.

discovery config schema will like this:

```yaml
dicovery: 
    type: object
    properties:
        service:
            type: object
            properties:
                schema:
                    type: string
                    enum: [ http ,https]
                    default:  https
                service_host:
                    type: string
                    default: ${KUBERNETES_SERVICE_HOST}  # format like ${XXX} means get value by os.getenv("xxx")
                service_port:
                    type: string
                    default: ${KUBERNETES_SERVICE_PORT}             
        client:
            type: object
            properties:
                client_token_data:
                    type: string                     
                client_token_file:
                    type: string
                    default: /var/run/secrets/kubernetes.io/serviceaccount/token
            oneOf:
            - required: [ client_token_data ]
            - required: [ client_token_file ]                   
        # namespaceSelector:       will supported later;                    
```

Reply:
@tokers , @tao12345666333  , please give some suggestions ,TKS